streams/src/main/java/org/apache/kafka/streams/processor/internals/SourceNode.java

```diff
@@ -28,28 +30,32 @@
     private final List<String> topics;
 
     private ProcessorContext context;
-    private Deserializer<K> keyDeserializer;
-    private Deserializer<V> valDeserializer;
+    private ExtendedDeserializer<K> keyDeserializer;
+    private ExtendedDeserializer<V> valDeserializer;
     private final TimestampExtractor timestampExtractor;
 
     public SourceNode(String name, List<String> topics, TimestampExtractor timestampExtractor, Deserializer<K> keyDeserializer, Deserializer<V> valDeserializer) {
         super(name);
         this.topics = topics;
         this.timestampExtractor = timestampExtractor;
-        this.keyDeserializer = keyDeserializer;
-        this.valDeserializer = valDeserializer;
+        this.keyDeserializer = keyDeserializer != null ? ensureExtended(keyDeserializer) : null;
+        this.valDeserializer = valDeserializer != null ? ensureExtended(valDeserializer) : null;
     }
 
     public SourceNode(String name, List<String> topics, Deserializer<K> keyDeserializer, Deserializer<V> valDeserializer) {
         this(name, topics, null, keyDeserializer, valDeserializer);
     }
 
-    K deserializeKey(String topic, byte[] data) {
-        return keyDeserializer.deserialize(topic, data);
+    private <T> ExtendedDeserializer<T> ensureExtended(Deserializer<T> deserializer) {
```

Comment:
Agreed, this can be moved to the Wrapper as a static method. I would suggest doing this not only for ExtendedDeserializer but also ExtendedSerializer so the pattern is the same.

```
public interface ExtendedSerializer<T> extends Serializer<T> {

    byte[] serialize(String topic, Headers headers, T data);

    class Wrapper<T> implements ExtendedSerializer<T> {

        private final Serializer<T> serializer;

        public Wrapper(Serializer<T> serializer) {
            this.serializer = serializer;
        }

        @Override
        public byte[] serialize(String topic, Headers headers, T data) {
            return serialize(topic, data);
        }

        @Override
        public void configure(Map<String, ?> configs, boolean isKey) {
            serializer.configure(configs, isKey);
        }

        @Override
        public byte[] serialize(String topic, T data) {
            return serializer.serialize(topic, data);
        }

        @Override
        public void close() {
            serializer.close();
        }

        public static <T> ExtendedSerializer<T> ensureExtended(Serializer<T> serializer) {
            return serializer == null ? null : serializer instanceof ExtendedSerializer ? (ExtendedSerializer<T>) serializer : new ExtendedSerializer.Wrapper<>(serializer);
        }
    }
}
```

```
public interface ExtendedDeserializer<T> extends Deserializer<T> {
    
    T deserialize(String topic, Headers headers, byte[] data);

    class Wrapper<T> implements ExtendedDeserializer<T> {

        private final Deserializer<T> deserializer;

        public Wrapper(Deserializer<T> deserializer) {
            this.deserializer = deserializer;
        }


        @Override
        public T deserialize(String topic, Headers headers, byte[] data) {
            return deserialize(topic, data);
        }

        @Override
        public void configure(Map<String, ?> configs, boolean isKey) {
            deserializer.configure(configs, isKey);
        }

        @Override
        public T deserialize(String topic, byte[] data) {
            return deserializer.deserialize(topic, data);
        }

        @Override
        public void close() {
            deserializer.close();
        }

        public static <T> ExtendedDeserializer<T> ensureExtended(Deserializer<T> deserializer) {
            return deserializer == null ?  null : deserializer instanceof ExtendedDeserializer ? (ExtendedDeserializer<T>) deserializer : new ExtendedDeserializer.Wrapper<>(deserializer);
        }
    }
}
```